File: sdk/typescript-guest/runtime/text.ts
/** The Unicode White_Space set used by the other maintained tutorial capsules. */
export function whitespace(value: string): boolean {
  return /^[\u0009-\u000D\u0020\u0085\u00A0\u1680\u2000-\u200A\u2028\u2029\u202F\u205F\u3000]$/u.test(value);
}
export function trim(value: string): string {
  const scalars = [...value];
  let start = 0, end = scalars.length;
  while (start < end && whitespace(scalars[start]!)) start++;
  while (end > start && whitespace(scalars[end - 1]!)) end--;
  return scalars.slice(start, end).join('');
}
export function utf8Length(value: string): number {
  let bytes = 0;
  for (const scalar of value) {
    const code = scalar.codePointAt(0)!;
    if (code >= 0xD800 && code <= 0xDFFF) throw new TypeError('unpaired UTF-16 surrogate');
    bytes += code < 0x80 ? 1 : code < 0x800 ? 2 : code < 0x10000 ? 3 : 4;
  }
  return bytes;
}

File: sdk/dotnet-guest/ownership/Text.cs
using System.Text;

namespace Lsf.Guest;

/// <summary>The tutorial contract uses Unicode White_Space, as Rust str::trim.</summary>
public static class Text
{
    public static bool Whitespace(char c) => c is >= '\x09' and <= '\x0d'
        or '\x20' or '\x85' or '\u00a0' or '\u1680'
        or >= '\u2000' and <= '\u200a' or '\u2028' or '\u2029'
        or '\u202f' or '\u205f' or '\u3000';
    public static string Trim(string value)
    {
        int begin = 0, end = value.Length;
        while (begin < end && Whitespace(value[begin])) ++begin;
        while (end > begin && Whitespace(value[end - 1])) --end;
        return value[begin..end];
    }
    public static int Utf8Length(string value) => Encoding.UTF8.GetByteCount(value);
}
